### PR TITLE
Add the bam_to_scidx tool

### PR DESCRIPTION
**Pull request by gregvonkuster**
which is a component of the soon-to-arrive Galaxy ChIP-exo flavor.

### CODE REVIEW

**Comment by gregvonkuster** [comment]
Related to: https://github.com/galaxyproject/tools-iuc/issues/404

File: tools/bam_to_scidx/bam_to_scidx.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+tmp_stdout = open(tmp_out, 'wb')
+tmp_err = tempfile.NamedTemporaryFile().name
+tmp_stderr = open(tmp_err, 'wb')
+
+# Link input BAM file and associated index file into the working directory.
+input_data_file_name = "input.bam"
+os.link(options.input_bam, os.path.join(tmp_dir, input_data_file_name))
+os.link(options.input_bai, os.path.join(tmp_dir, "%s.bai" % input_data_file_name))
+
+cmd = "java -jar %s -b %s -s %d -o %s" % (options.jar_file, input_data_file_name, options.read, options.output)
```
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.py]: Is there any reason you can not simply call this commandline from within cheetah? Linking could be done by plain `ln -s` and all the subprocess and argparse magic would be redundant.

File: tools/bam_to_scidx/bam_to_scidx.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+the Chip-exo Galaxy flavor.  ScIdx files are 1-based.  The format consists of 5 columns: the chromosome,
+the position of the genomic coordinate, the number of tags on the forward strand, the number of tags on
+the reverse strand and the number of total tags on the position.  With pair-end reads, only the 5’ end of
+READ1 will be used to create the ScIdx data file.  Tools that use this format include GeneTrack and MultiGPS.
+
+    </help>
+    <citations>
+        <citation type="bibtex">
+            @unpublished{None,
+            author = {None},
```
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.xml]: This looks strange. You know the authors right :)

File: tools/bam_to_scidx/bam_to_scidx.xml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    <inputs>
+        <param name="input_bam" type="data" format="bam" label="BAM file" help="BAM file must be sorted and indexed" />
+        <param name="read" type="select" label="Read to output">
+            <option value="0" selected="True">Read1</option>
+            <option value="1">Read2</option>
+            <option value="2">Combined</option>
+        </param>
+    </inputs>
+    <outputs>
+        <data name="output" format="scidx" label="${tool.name} on ${on_string}" />
```
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.xml]: I guess the label can be removed if you use this default label.

File: tools/bam_to_scidx/bam_to_scidx.xml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            python $__tool_directory__/bam_to_scidx.py
+            -j $__tool_directory__/BAMtoIDX.jar
+            -b "$input_bam"
+            -i "${input_bam.metadata.bam_index}"
+            -r $read
+            -o "$output"
+        ]]>
+    </command>
+    <inputs>
+        <param name="input_bam" type="data" format="bam" label="BAM file" help="BAM file must be sorted and indexed" />
```
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.xml]: The help is redundant as every default BAM dataset is sorted and indexed. Or do you mean name-sorted?

File: tools/bam_to_scidx/bam_to_idx.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
+import htsjdk.samtools.AbstractBAMFileIndex;
```
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_idx.java]: Mh, this raises an interesting point. Do we want to have java source files included if we just use the jar file?
Anyone?

File: tools/bam_to_scidx/bam_to_scidx.py (hunk 1 above)
Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.py]: The main reason is that the Java component of this tool (and in some cases Java itself) spew out stuff to stdout which I've captured and eliminated using the Python script. ;)

File: tools/bam_to_scidx/bam_to_scidx.xml (hunk 2 above)
Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.xml]: Yes, I can add a name, but didn't know for sure if the tool developer's name should go here.  I'll add it.

File: tools/bam_to_scidx/bam_to_scidx.py (hunk 1 above)
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.py]: Can you not pipe it to /dev/null ? I also don't see a problem to write to stdout. Picard does this as well.

Comment (nsoranzo) [inline tools/bam_to_scidx/bam_to_scidx.py]: @gregvonkuster `1>/dev/null` ?

Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.py]: Ok, working on it.  Thanks!

Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.py]: @nsoranzo @bgruening Are there any examples of tools that only have a config that create symlinks in the command tag set for the Bam Indexes?  Thanks!

Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.py]: https://github.com/galaxyproject/tools-iuc/blob/master/tools/bedtools/multiCov.xml#L11 :)

File: tools/bam_to_scidx/bam_to_scidx.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+tmp_dir = tempfile.mkdtemp(prefix='tmp-scidx-')
+tmp_out = tempfile.NamedTemporaryFile().name
+tmp_stdout = open(tmp_out, 'wb')
+tmp_err = tempfile.NamedTemporaryFile().name
+tmp_stderr = open(tmp_err, 'wb')
+
+# Link input BAM file and associated index file into the working directory.
+input_data_file_name = "input.bam"
+os.link(options.input_bam, os.path.join(tmp_dir, input_data_file_name))
+os.link(options.input_bai, os.path.join(tmp_dir, "%s.bai" % input_data_file_name))
```
Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.py]: @bgruening @nsoranzo I'm kinda stumped on how to replace the above 2 lines in the tool config command tag set.  ;(
Maybe something like:
ln -s "${__job_working_directory__/input_bam.metadata.bam_index} input.bam

I've not seen this syntax in any other tools...

File: tools/bam_to_scidx/bam_to_scidx.py (hunk 1 above)
Comment (nsoranzo) [inline tools/bam_to_scidx/bam_to_scidx.py]: @gregvonkuster Just remember to connect the 2 commands with `&&` .

File: tools/bam_to_scidx/bam_to_scidx.py (hunk 6 above)
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.py]: ln -s $inputbam.metadata.bam_index foo.bam.bai &&

Comment (yhoogstrate) [inline tools/bam_to_scidx/bam_to_scidx.py]: `ln -f -s "${input_bam.metadata.bam_index}" "${input_bam}.bai" &&` :wink:

**Comment by gregvonkuster** [comment]
@bgruening @nsoranzo @yhoogstrate Thanks for the review and help.  I think I've captured all of the feedback in the latest commit.  Please let me know if I've missed something.  :)

File: tools/bam_to_scidx/bam_to_idx.java (hunk 5 above)
Comment (erasche) [inline tools/bam_to_scidx/bam_to_idx.java]: No, I don't think so. No different than a compiled binary like samtools.

Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_idx.java]: @erasche @bgruening I can certainly delete it.  My thought with including it is that the tool is very specific, and the developer will more easily be able to maintain the code if the source is here.  I suppose the Java source could be placed in a separate github repo though.

**Comment by bgruening** [comment]
Looks so much simpler now. Thanks @gregvonkuster!

**Comment by gregvonkuster** [comment]
Thanks @bgruening I'm still not quire sure how the citation should look since there is no publication yet.  I remember an email thread form way back where I believe it was agreed that just a simple URL will suffice, so that is why I have it this way.  If you want it changed, please let me know.

**Comment by bgruening** [comment]
If it renders nicely I'm ok with that :)

File: tools/bam_to_scidx/bam_to_idx.java (hunk 5 above)
Comment (erasche) [inline tools/bam_to_scidx/bam_to_idx.java]: @gregvonkuster the case you describe, I'm fine with tool code in repo. No different than the python wrappers the rest of us write.

**Comment by gregvonkuster** [comment]
Just added a WIP tag to this - I will be applying an enhancement to this wrapper today.

**Comment by gregvonkuster** [comment]
@bgruening @erasche Sorry about moving this back to a WIP after initial review.  The updated tool just includes 3 new, additional optional parameters, and the jar file has been renamed to conform more closely to the tool name.  The underlying behavior of the tool is the same.

File: tools/bam_to_scidx/bam_to_scidx.xml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        <param name="require_proper_mate_pairing" type="select" label="Require proper mate-pairing?" help="Required if filtering by insert size for single-end Reqd 1 (below)." >
+            <option value="1" selected="True">Yes</option>
+            <option value="0">No</option>
+        </param>
+        <param name="read" type="select" label="Read to output">
+            <option value="0" selected="True">Read1</option>
+            <option value="1">Read2</option>
+            <option value="2">Combined</option>
+        </param>
+        <param name="min_insert_size" type="integer" value="" optional="True" min="0" label="Minimum insert size to output" help="Will not filter out single-end Read 1 unless proper mate-pairing is required (above)."/>
```
Comment (bgruening) [inline tools/bam_to_scidx/bam_to_scidx.xml]: Can you please add -m -M and above -p to the help attribute. Thanks!

Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.xml]: @bgruening Done - thanks!  ;)

Comment (gregvonkuster) [inline tools/bam_to_scidx/bam_to_scidx.xml]: Ping @bgruening @martenson @erasche @nsoranzo can this tool be merged?  Let me know if there is more I should do.  Thanks!

**Comment by gregvonkuster** [comment]
@erasche Thanks!  ;)

**Comment by erasche** [comment]
No problem. Meant to get around to earlier. Will put on MTS when I get home

**Comment by gregvonkuster** [comment]
@erasche I have this on the TTS owned by greg, but I will deprecate it when it is available owned by iuc.  Thanks!

**Comment by bgruening** [comment]
It's uploaded to both TS.

**Comment by gregvonkuster** [comment]
@bgruening Thanks!  I've deprecated the one owned by me on TTS.  Much appreciated!

**Comment by bgruening** [comment]
Thanks for you work and contribution to this repo!!!!

**Pull request merged**